File: leetcode/combination_sum/combination_sum.cpp
/*
 * Time Complexity: O(2^T * K)
 *   - T is the target value, K is the average length of each combination.
 *   - In the worst case, the number of combinations is exponential in T.
 *
 * Space Complexity: O(T)
 *   - The recursion stack can go as deep as T.
 *   - Additional space for storing the answer.
 */

#include<vector>
#include<algorithm>

using std::vector;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> currVec;
        sort(candidates.begin(), candidates.end());

        currVec.push_back(candidates[0]);
        combinations(ans, candidates, currVec, target, 0, candidates[0]);
        currVec.pop_back();
        combinations(ans, candidates, currVec, target, 1, 0);

        return ans;
    }

    bool combinations(
        vector<vector<int>>& ans,
        vector<int>& candidates,
        vector<int>& currVec,
        int target,
        int currIndex,
        int currSum
    ) {
        if(currSum == target) {
            ans.push_back(vector<int>(currVec));
            return true;
        }

        if(currIndex >= candidates.size()) return false;

        int tmpCurrSum = currSum + candidates[currIndex];
        if(tmpCurrSum > target) return false;

        currVec.push_back(candidates[currIndex]);
        bool first = combinations(ans, candidates, currVec, target, currIndex, tmpCurrSum);
        currVec.pop_back();
        if(!first) combinations(ans, candidates, currVec, target, currIndex+1, currSum);

        return false;
    }
};

File: leetcode/combination_sum/combination_sum.java
package combination_sum;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

class Solution {
    private int[] candidates;
    private List<List<Integer>> ans;
    private List<Integer> currAns;
    private int target;
    
    private boolean combination(int currSum, int currIndex) {
        if(currSum == target) {
            ans.add(new ArrayList<>(currAns));
            return true;
        }

        if(currIndex >= candidates.length) return false;

        int tmpCurrSum = currSum + candidates[currIndex];
        if(tmpCurrSum > target) return false;

        currAns.add(candidates[currIndex]);
        boolean first = combination(tmpCurrSum, currIndex);
        currAns.remove(currAns.size()-1);
        if(!first) combination(currSum, currIndex+1);

        return false;
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        this.candidates = candidates;
        this.target = target;
        this.ans = new LinkedList<>();
        this.currAns = new ArrayList<>();

        Arrays.sort(candidates);

        currAns.add(candidates[0]);
        combination(candidates[0], 0);
        currAns.remove(currAns.size()-1);
        combination(0, 1);

        return ans;
    }


}